### PR TITLE
support gw single limit 

### PR DESCRIPTION
**Pull request by MouceL**
目前网关网格需要分开部署，针对网关场景：

设置 disableInsertLocalRateLimit =true,  并 apply 形似以下EnvoyFilter 至网关pilot
```
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  labels:
    istio.io/rev:  your_gateway
  name: envoyfilters-gateway
  namespace: gateway
spec:
  configPatches:
    - applyTo: HTTP_FILTER
      match:
        context: GATEWAY
        listener:
          filterChain:
            filter:
              name: envoy.filters.network.http_connection_manager
              subFilter:
                name: envoy.filters.http.router
      patch:
        operation: INSERT_BEFORE
        value:
          name: envoy.filters.http.local_ratelimit
          typed_config:
            '@type': type.googleapis.com/udpa.type.v1.TypedStruct
            type_url: type.googleapis.com/envoy.extensions.filters.http.local_ratelimit.v3.LocalRateLimit
            value:
              stat_prefix: http_local_rate_limiter
```

之后创建  SmartLimiter

```
apiVersion: microservice.slime.io/v1alpha2
kind: SmartLimiter
metadata:
  name: prod-gateway
  namespace: gateway-v112
spec:
  gateway: true  
  sets:
    _base:
      descriptor:
      - action:
          fill_interval:
            seconds: 60
          quota: "3"
          strategy: single
        condition: "true"
        target:
          direction: outbound
          route:
          - a.test.com:80/r1
  workloadSelector:
    gw_cluster: prod-gateway   
```

smartlimiter 生成的 envoyfilter 如下 
```
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: prod-gateway.gateway-v112.ratelimit
  namespace: gateway-v112
  ownerReferences:
  - apiVersion: microservice.slime.io/v1alpha2
    kind: SmartLimiter
    name: prod-gateway
spec:
  configPatches:
  - applyTo: HTTP_ROUTE
    match:
      context: GATEWAY
      routeConfiguration:
        vhost:
          name: a.test.com:80
          route:
            name: r1
    patch:
      operation: MERGE
      value:
        route:
          rate_limits:
          - actions:
            - generic_key:
                descriptor_value: Service[prod-gateway.gateway-v112]-User[none]-Id[4031918882]
  - applyTo: HTTP_ROUTE
    match:
      context: GATEWAY
      routeConfiguration:
        vhost:
          name: a.test.com:80
          route:
            name: r1
    patch:
      operation: MERGE
      value:
        typed_per_filter_config:
          envoy.filters.http.local_ratelimit:
            '@type': type.googleapis.com/udpa.type.v1.TypedStruct
            type_url: type.googleapis.com/envoy.extensions.filters.http.local_ratelimit.v3.LocalRateLimit
            value:
              descriptors:
              - entries:
                - key: generic_key
                  value: Service[prod-gateway.gateway-v112]-User[none]-Id[4031918882]
                token_bucket:
                  fill_interval:
                    seconds: "60"
                  max_tokens: 3
                  tokens_per_fill:
                    value: 3
              filter_enabled:
                default_value:
                  numerator: 100
                runtime_key: local_rate_limit_enabled
              filter_enforced:
                default_value:
                  numerator: 100
                runtime_key: local_rate_limit_enforced
              stat_prefix: http_local_rate_limiter
              token_bucket:
                fill_interval:
                  seconds: "1"
                max_tokens: 100000
                tokens_per_fill:
                  value: 100000
  workloadSelector:
    labels:
      gw_cluster: prod-gateway  
```

### CODE REVIEW

**Comment by YonkaFang** [comment]
it seems that it may cause gw hcm have multiple`type.googleapis.com/envoy.extensions.filters.http.local_ratelimit.v3.LocalRateLimit` filters.

**Comment by MouceL** [comment]
> the `context: GATEWAY` does not work for `HTTP_FILTER` patch?

it work well when set to  `context: GATEWAY` 

but no match will take effect in inbound/outbound/gateway  well

**Comment by YonkaFang** [comment]
> 

we should be more rigorous as well as more ideological

**Comment by YonkaFang** [comment]
also we should add more `Gateway`-semantic api(fields) later

**Comment by MouceL** [comment]
add disableInsertLocalRateLimit , and you can set disableInsertLocalRateLimit=true, and apply envoyfilter like yaml in staging/src/slime.io/slime/modules/limiter/install/gw_limiter_envoyfilter.yaml to your gw pilot

File: staging/src/slime.io/slime/modules/limiter/controllers/envoy_filters.go [hunk 1]
```diff
@@ -143,10 +156,15 @@ func descriptorsToEnvoyFilter(descriptors []*microservicev1alpha2.SmartLimitDesc
 
 	// enable and config plugin envoy.filters.http.local_ratelimit
 	if len(localDescriptors) > 0 {
-		httpFilterLocalRateLimitPatch := generateHttpFilterLocalRateLimitPatch()
-		ef.ConfigPatches = append(ef.ConfigPatches, httpFilterLocalRateLimitPatch)
-
-		perFilterPatch := generateLocalRateLimitPerFilterPatch(localDescriptors, loc)
+		// TODO if disableInsertLocalRateLimit=false in gw, multi smartlimiter will patch duplicate configurations, This will lead dysfunctional behavior.
```
Comment (cywang1905) [inline staging/src/slime.io/slime/modules/limiter/controllers/envoy_filters.go:159]: Seems nothing todo, or you want to improve this solution in the future? Then intro the idea more clearly.

File: staging/src/slime.io/slime/modules/limiter/controllers/envoy_local_limit.go [hunk 2]
```diff
@@ -55,12 +57,14 @@ func generateRouteConfigs(target *microservicev1alpha2.SmartLimitDescriptor_Targ
...
 				vhostName: vhostName,
 				direction: target.Direction,
+				gw:        gw,
 			}
 			rcs = append(rcs, rc)
 		}
 		return rcs
 	}
 	// build inbound by default
+	log.Infof("build default inbound route config base on %s", target)
```
Comment (cywang1905) [inline staging/src/slime.io/slime/modules/limiter/controllers/envoy_local_limit.go:67]: The info level is fine if you do not print too many repeated detail logs during normal operation. You can have a think about if you need to change the log level according to the actual situation.

File: staging/src/slime.io/slime/modules/limiter/controllers/smartlimiter_controller.go [hunk 3]
```diff
@@ -198,6 +199,79 @@ func newProducerConfig(env bootstrap.Environment) (*metric.ProducerConfig, error
...
+
+func (r *SmartLimiterReconciler) RemoveInterested(req ctrl.Request) {
+	key := FQN(req.Namespace, req.Name)
+	r.metricInfo.Pop(key)
+	r.interest.Pop(key)
+	log.Infof("name %s, namespace %s has poped", req.Name, req.Namespace)
+
+	// if contain global smart limiter, should delete info in configmap
+	if !r.cfg.GetDisableGlobalRateLimit() {
+		log.Infof(" refresh global rate limiter configmap")
```
Comment (cywang1905) [inline staging/src/slime.io/slime/modules/limiter/controllers/smartlimiter_controller.go:228]: unnecessary blank

File: staging/src/slime.io/slime/modules/limiter/install/gw_limiter_envoyfilter.yaml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+                name: envoy.filters.http.router
+      patch:
+        operation: INSERT_BEFORE
+        value:
+          name: envoy.filters.http.local_ratelimit
+          typed_config:
+            '@type': type.googleapis.com/udpa.type.v1.TypedStruct
+            type_url: type.googleapis.com/envoy.extensions.filters.http.local_ratelimit.v3.LocalRateLimit
+            value:
+              stat_prefix: http_local_rate_limiter
```
Comment (cywang1905) [inline staging/src/slime.io/slime/modules/limiter/install/gw_limiter_envoyfilter.yaml:27]: Add a blank line at the end

**Review by cywang1905 — APPROVED** [review]

**Pull request merged**